cmd/dockerd/daemon.go

```diff
@@ -228,6 +232,35 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 	// Notify that the API is active, but before daemon is set up.
 	preNotifyReady()
 
+	// Setup buildkit trace recorder
+	detect.Recorder = detect.NewTraceRecorder()
```

Comment:
Not sure this is the best place to put this (to me it can be left in the init as it is just a variable definition and no slow path).

Afaics, there are now 2 separate trace providers, one coming from `getTracerProvider()` and the other from `detect.TraceProvider()`. The second one will handle only the requests coming to the `/grpc` and only one that buildkit sees. This may be fine, but in that case it makes little sense to initialize `detect.Recorder` here together with `getTracerProvider` because they don't have any shared parts. Technically I think the current code works as the first code calling `detect.TraceProvider()` is in `NewRouter`, so called after this codepath.

Reply:
I changed the grpc router to just read the tracer provider from the otel lib.
I definitely prefer to have the recorder set here where we control ordering rather than in an init.

Reply:
I don't understand the latest changes. If the requests coming in from the grpc interceptors use a different trace provider this will mean that the spans will not appear in buildkit traces as well as that initialization of `Recorder` does not do anything.

My comment about the location on recorder initialization was that it's not related to `getTracerProvider()` and although the current naming seems to suggest it.

Reply:
Ok, I'm stepping through buildkit to see how this is used and see the problem now.
`detect.TracerProvider()` is using a new tracer provider to wrap any configured span exporter with the recorder.

How can we test this so it doesn't break?

Reply:
In buildctl you can do:
```
buildctl debug histories --format '{{json .}}'
// find a build and a trace record
buildctl debug get sha256:db7dbd0f7d0d2239075f83ec6b2dd76bd7d9558bdb79895242c7a3944c6dadaf
```

If you have Docker Desktop installed then alternative:
```
curl --unix-socket ~/Library/Containers/com.docker.docker/Data/docker-desktop-build.sock http://localhost/builds/<buildx builder>/<buildx node>/<build ref>/stream
// get trace
curl --unix-socket ~/Library/Containers/com.docker.docker/Data/docker-desktop-build.sock 'http://localhost/blobs/<buildx builder>/<buildx node>?digest=sha256:8727e058fe0ed45dac529e8c8390107a001e545d30a83c6c148a2aa8d5058df9'
```

Reply:
Ok, I've added a test case for this.
The reason this was taking me so long is not realizing the trace data gets set on the history record async (which of course makes sense) so it looked like it wasn't populating.
Validated the test fails when this was setup incorrectly.

Also validated buildkit client and server traces are exported with the rest of the daemon traces.
<img width="858" alt="image" src="https://github.com/moby/moby/assets/799078/e9ee20a4-b730-4213-9693-6bcc46e1918b">